crates/wast/src/core/binary.rs

```diff
@@ -174,10 +174,22 @@ impl Encode for RecOrType<'_> {
 impl Encode for Type<'_> {
     fn encode(&self, e: &mut Vec<u8>) {
         if let Some(parent) = &self.parent {
-            e.push(0x50);
+            if self.final_type.is_some() && self.final_type.unwrap() == true {
+                e.push(0x4e);
+            }
+            else {
+                e.push(0x50);
+            }
             (1 as usize).encode(e);
             parent.encode(e);
         }
+        else {
+            // No supertype but sub was used
+            if self.final_type.is_some() && self.final_type.unwrap() == false {
+                e.push(0x50);
+                (0 as usize).encode(e);
+            }
+        }
```

Comment:
To simplify this a bit, could it perhaps become:

```rust
match (&self.parent, self.final_type) {
    (Some(parent), Some(true)) => {
        e.push(0x4e);
        e.push(0x01); // same as `(1 as usize).encode(e)` I think
        parent.encode(e);
    }
    (Some(parent), Some(false) | None) => {
        e.push(0x50);
        e.push(0x01);
        parent.encode(e);
    }
    (None, Some(false)) => {
        e.push(0x50);
        e.push(0x00);
    }
    (None, _) => {} // comment about why no prefix is necessary
}
```

Or.. well after I type this out I'm less certain. I don't know anything much about the gc spec but I'm not sure I fully understand this encoding scheme, so I'm also happy to defer to you.

Reply:
Thanks, it's definitely easier to understand that way! I rebased with your suggestion and some additional comments.